core/src/main/java/org/elasticsearch/rest/RestController.java

```diff
@@ -242,22 +243,39 @@ private boolean hasContentTypeOrCanAutoDetect(final RestRequest restRequest, fin
                 // be removed!
                 deprecationLogger.deprecated("Plain text request bodies are deprecated. Use request parameters or body " +
                     "in a supported format.");
-            } else if (isContentTypeRequired) {
-                return false;
-            } else {
-                deprecationLogger.deprecated("Content type detection for rest requests is deprecated. Specify the content type using " +
-                    "the [Content-Type] header.");
-                XContentType xContentType = XContentFactory.xContentType(restRequest.content());
-                if (xContentType == null) {
+            } else if (restHandler != null && restHandler.supportsContentStream() && restRequest.header("Content-Type") != null) {
+                final String lowercaseMediaType = restRequest.header("Content-Type").toLowerCase(Locale.ROOT);
```

Comment:
Now that we have this new if, we could also better validate the content-type accepted by these apis, cause I think when providing yaml or cbor you get back an NPE.